tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,460 @@
+/**
+ * @file   mem_filesystem.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2020 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements the in-memory filesystem class
+ */
+
+#include <sstream>
+#include <unordered_set>
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/mem_filesystem.h"
+
+using namespace tiledb::common;
+using namespace std;
```

Comment:
We typically omit this as we may use objects like `vector` and `array` in our own namespace. Just a minor remark.

Reply:
I agree this is not a good practice, but I think @joe-maley mentioned some suggestion about using it only in .cc files that was discussed with @Shelnutt2 ?

Reply:
Stavros brings up a good point that there are similarly named classes in `std`. Earlier, I was referring to a discussion that @stavrospapadopoulos and I had where we decided it would be ok to use `using namespace tiledb::common` in .cc files, which I assumed set the precedence for similarly using `using namespace std`. Let's discuss this offline to reach a consensus.

Reply:
@ypatia We're sticking with disallowing `using namespace std`, but `using namespace tiledb::...` is still allowed. Sorry for the confusion!

Reply:
sure, no problem, I am removing the line. Thanks everyone for the discussion and clarifications